<a href="https://colab.research.google.com/github/umi-na/google_colab/blob/main/topic_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 토픽 #3: 데이터 수집 - 데이터 읽기/병합/제거

Colab 실습 시작 전, 이 노트북의 **복사본을 자신의 구글 드라이브에 저장**함: Colab 메뉴에서 "File" >> "Save a Copy in Drive". "Save a Copy" 메뉴 선택 시 "Copy of topic#3.ipynb" 복사본이 구글 드라이브의 'Colab Notebook' 폴더에 저장됨. 그리고 본사본이 새 브라우저 창에 열림.

# 데이터 수집
![CSV Screenshot](https://drive.google.com/uc?export=view&id=18oV5qlzX63XJonj9HNJjn4FWg-9yCKt-) 

# 수업 목표
* 데이터 읽기

* 데이터 병합

* 데이터 삭제

# 데이터 읽기
* 데이터 수집 단계에서 "데이터 읽기/쓰기"는 가장 기본/기초 과정 중 한개임

* Pandas라이브러리를 활용하여 {CSV, Excel, JSON, HTML} 파일을 읽는 법에 대해 배우겠음

* 이해들 돋기 위해 타이타닉 탑습객 데이터를 활용하여 파일 읽기에 대해 배우겠음

## 타이타닉 탑습객 - 예시 데이터 소개

* [Kaggle](https://www.kaggle.com/)의 [타이타닉 탁습객 데이터](https://www.kaggle.com/c/titanic/data)를 활용해서 Pandas라이브러리로 읽어오는 법을 배우겠음

* 본 수업을 위해 탑습객 데이터를 조금 수정하여 {CSV, Excel, JSON}파일로 변형하였음

* 본 수업에서는 변형된 파일을 사용함([변형된 파일 다운로드 링크](https://drive.google.com/drive/folders/1KpfVquLkXBRByHmpwCvxHYpbpj7sebki?usp=sharing))

* 타이타닉 탁습객 데이터는 아래와 같은 정보가 기록되어 있음

컬럼 이름|	설명|	컬럼값 설명
---| ---| ---
PassengerId| 탑승객 번호| 
Survived|	생존유무|	0 = 사망, 1 = 생존
pclass|	티켓 글래스|	1 = 1등석, 2 = 2등석, 3 = 3등석
Sex|	성별| Female = 여성, Male = 남성	
Age|	만 나이 | 태어난지 1년 미만으면(만 나이 < 0) 소수로 표현
Ticket|	티켓 번호|	
Fare|	티켓 가격|	
Cabin|	객실 번호|	
Embarked|	탑승 항구|	C = 셰르부르(프랑스 서북부 항구), Q = 퀸즈타운(아일랜드 남부의 항구), S = 사우샘프턴 (영국 남부 항구)

* 변형된 데이터 파일을 다운받아 구글 드라이브에 업로드하여 사용하면 됨

## 구글 드라이브 폴더 경로 설정



* 아래 root_path의 스트링값은 파일이 있는 구글 드라이브 폴더 경로를 나타냄 

* 자신에게 맞게 구글 드라이브 폴더 경로를 변경하여 사용하면 됨

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
root_path = '/content/drive/MyDrive/Colab Notebooks/Colab/wk3_data'

## CSV 파일 읽기/쓰기

### CSV 파일 읽기

![CSV Screenshot](https://drive.google.com/uc?export=view&id=1Z8K_TSH-K7DrPBEqi06VbhhCc5gkGv_Q) 

* [pandas.read_csv](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) 메소드를 사용하여 CSV 파일을 DataFrame으로 읽어올 수 있음

* pandas.read_csv 메소드에 파일경로를 입력값으로 넣으면 됨

* 아래 코드는 메소드를 사용하여 타이타닉 탑승객 정보 CSV 파일 (i.e., '*titanic.csv*')을 읽음

* [os.path.join](https://docs.python.org/3/library/os.path.html)(경로#1, 경로#2...) 메소드는 입력된 모든 경로를 합쳐 한개의 경로로 만들어 반환함

In [ ]:
path.join(root_path,'titanic.csv') #root_path+'titanic.csv'를 사용해도 되긴 하지만 os별로 /가 어떻게 붙을지(//나 \로 붙는 경우가 있다) 모르기때문에 path.join을 써주는 것이 좋다

'/content/drive/MyDrive/wk3_data/titanic.csv'

In [ ]:
import pandas as pd
from os import path

df = pd.read_csv(path.join(root_path,'titanic.csv'))
df.head() #어떤 DataFrame의 첫 다섯 개의 정보를 보여줌

PassengerId  Survived  Pclass  ...     Fare  Cabin Embarked
0            1         0       3  ...   7.2500    NaN        S
1            2         1       1  ...  71.2833    C85        C
2            3         1       3  ...   7.9250    NaN        S
3            4         1       1  ...  53.1000   C123        S
4            5         0       3  ...   8.0500    NaN        S

[5 rows x 9 columns]

* CSV 데이터의 구분자(separator)가 쉼표(',')가 아니면 구분자 설정이 필요함

* 아래 그림은 '|'을 구분자로 사용하는 CSV 데이터('*titanic_sep.csv*') 예시임

![CSV Screenshot](https://drive.google.com/uc?export=view&id=1jJn2z65CUs7gHtpRkXeRtFTQTYmGWL90)

* pandas.read_csv 메소드의 sep 변수를 활용하여 구분자를 설정하여 CSV파일을 읽어 올 수 있음

* 먼저 '|'을 구분자로 사용하는 CSV 데이터('*titanic_sep.csv*')를 sep 변수 설정 없이 읽어 보겠음

* 총 9개의 컬럼이 있지만 DataFrame에 컬럼이 한개만 생성된 것을 알 수 있음


In [ ]:
df = pd.read_csv(path.join(root_path,'titanic_sep.csv'))
df.head()

PassengerId|Survived|Pclass|Sex|Age|Ticket|Fare|Cabin|Embarked
0                    1|0|3|male|22|A/5 21171|7.25||S            
1             2|1|1|female|38|PC 17599|71.2833|C85|C            
2          3|1|3|female|26|STON/O2. 3101282|7.925||S            
3                 4|1|1|female|35|113803|53.1|C123|S            
4                       5|0|3|male|35|373450|8.05||S



* 다음으로 sep 변수를 '|'로 설정하여 읽어 보겠음. 9개의 컬럼이 잘 생성된 것을 알 수 있음

In [ ]:
df = pd.read_csv(path.join(root_path,'titanic_sep.csv'), sep='|') #sep을 사용하여 |를 구분자로 사용함
df.head()

PassengerId  Survived  Pclass  ...     Fare  Cabin Embarked
0            1         0       3  ...   7.2500    NaN        S
1            2         1       1  ...  71.2833    C85        C
2            3         1       3  ...   7.9250    NaN        S
3            4         1       1  ...  53.1000   C123        S
4            5         0       3  ...   8.0500    NaN        S

[5 rows x 9 columns]

### CSV 파일 쓰기

* [DataFrame.to_csv](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_csv.html) 메소드로 DataFrame을 CSV파일로 저장할 수 있음

* 아래 코드는 df1, df2를 CSV파일로 저장함

In [ ]:
df1 = pd.DataFrame({'A': ['df1_A0', 'df1_A1', 'df1_A2'],
                    'B': ['df1_B0', 'df1_B1', 'df1_B2'],
                    'C': ['df1_C0', 'df1_C1', 'df1_C2']},
                   index=[0, 1, 2])
df2 = pd.DataFrame({'A': ['df2_A2', 'df2_A3', 'df2_A4'],
                    'B': ['df2_B2', 'df2_B3', 'df2_B4'],
                    'C': ['df2_C2', 'df2_C3', 'df2_C4']},
                   index=[2, 3, 4])

* DataFrame.to_csv 메소드에 파일 경로(e.g., '*.../df1.csv*')를 넣으면 됨

In [ ]:
df1.to_csv(path.join(root_path,'df1.csv'))
df2.to_csv(path.join(root_path,'df2.csv'))

* 오른쪽 3번째 탭으로 저장된 CSV 파일(i.e., df1.csv, df2.csv) 확인 가능

![Excel Screenshot](https://drive.google.com/uc?export=view&id=1NmU1iT31VKS6261_TC8PionBdBIPZkWT) 


* 오른쪽 3번째 탭으로 파일 이름 더블 클릭 시 왼쪽 탭에 파일 데이터를 표시함. 데이터를 표시할 수 없는 파일은 다운로드됨
* CSV 파일(e.g., df1.csv) 클릭 시 왼쪽 탭에 파일 데이터를 테이블로 표시함

![Excel Screenshot](https://drive.google.com/uc?export=view&id=1Pw40anRF-X1OkWzBsxw6tn5_wfIxHHPO) 

## Excel 파일 읽기

### 엑셀 읽기 (기본)
![Excel Screenshot](https://drive.google.com/uc?export=view&id=1W03Va3J6wwq2wrC2iT8zFHm1etGHh5oF) 

* [pandas.read_excel](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_excel.html) 메소드를 사용하여 excel 파일을 DataFrame으로 읽어올 수 있음

* pandas.read_excel 메소드에 파일경로를 입력값으로 넣으면 됨

* 아래 코드는 메소드를 사용하여 타이타닉 탑승객 정보 excel 파일 (i.e., '*titanic.xlsx*')을 읽음

In [ ]:
df = pd.read_excel(path.join(root_path,'titanic.xlsx'))
df.head()

PassengerId  Survived  Pclass  ...      Fare Cabin  Embarked
0            2         1       1  ...   71.2833   C85         C
1            4         1       1  ...   53.1000  C123         S
2           12         1       1  ...   26.5500  C103         S
3           32         1       1  ...  146.5208   B78         C
4           53         1       1  ...   76.7292   D33         C

[5 rows x 12 columns]

### 시트 선택

* 엑셀 파일에 시트 페이지가 2개 이상 있을 수 있음. 이 경우 첫 페지 외 다른 페이지를 읽고 오고 싶으면 추가 설정이 필요함

* 시트 페이지 번호 또는 시트 이름을 pandas.read_excel 메소드에 설정해야 함

* 시트 페이지 번호 또는 시트 이름을 설정하는 법을 배우겠음

#### 번호로 선택

![Excel Screenshot](https://drive.google.com/uc?export=view&id=1HzPhs0TbZv_Eovb1zHE797QGEhhbRFCL)

* 윗 그림은 각 시트 페이지에 대응되는 시트 번호를 보려주고 있음

* 시트 번호는 0으로 시작함. 예컨대 총 3 페이지가 있으면 마지막 시트 번호는 2임

* 아래 코드는 Pandas.read_excel의 sheet_name을 '1'로 설정하여 시트번호 2번째 페이지(시트 번호 1)를 읽음

In [ ]:
df = pd.read_excel(path.join(root_path,'titanic.xlsx'), sheet_name=1) #두번째 시트의 정보를 가져옴
df.head()

PassengerId  Survived  Pclass     Sex   Age  Ticket     Fare Cabin Embarked
0           10         1       2  female  14.0  237736  30.0708   NaN        C
1           16         1       2  female  55.0  248706  16.0000   NaN        S
2           18         1       2    male   NaN  244373  13.0000   NaN        S
3           21         0       2    male  35.0  239865  26.0000   NaN        S
4           22         1       2    male  34.0  248698  13.0000   D56        S

#### 이름으로 선택
![CSV Screenshot](https://drive.google.com/uc?export=view&id=1_O63OdH0V6KAcxq7yNI9qMAOgkXfrK_r) 

* sheet_name에 실제 페이지 이름을 설정하여 해당 페이지를 읽어 올 수 있음

* 윗 그림은 각 시트 페이지의 시트 이름을 보여주고 있음

* 아래 코드는 Pandas.read_excel의 sheet_name을 'Pclass3'으로 설정하여 3번째 페이지를 읽음


In [ ]:
df = pd.read_excel(path.join(root_path,'titanic.xlsx'), sheet_name='Pclass3') #시트 이름으로 가져옴
df.head()

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            3         1       3  ...   7.9250   NaN         S
2            5         0       3  ...   8.0500   NaN         S
3            6         0       3  ...   8.4583   NaN         Q
4            8         0       3  ...  21.0750   NaN         S

[5 rows x 12 columns]

## JSON 파일 읽기
![CSV Screenshot](https://drive.google.com/uc?export=view&id=1_HC_1KIMi2iTJXeeVqTwkHrE77mHCsE5) 

* CSV, Excel 파일 외 JSON, HTML, SQL, SPSS 등의 다른 유형의 파일도 읽어와 DataFrame으로 생성 할 수 있음

* 아래 코드는 pandas.to_json 메소드로 JSON 파일('*titanic.json*')을 읽어와 DataFrame을 생성하는 예시 코드임


In [ ]:
df = pd.read_json(path.join(root_path,'titanic.json'))
df.head()

PassengerId  Survived  Pclass  ...     Fare  Cabin Embarked
0            1         0       3  ...   7.2500   None        S
1            2         1       1  ...  71.2833    C85        C
2            3         1       3  ...   7.9250   None        S
3            4         1       1  ...  53.1000   C123        S
4            5         0       3  ...   8.0500   None        S

[5 rows x 9 columns]

## 다른 유형의 파일 읽기

* 이전 코드와 유사하게 다른 유형의 파일도 읽을 수 있음.

* 데이터 파일을 읽을 때 사용하는 메소드의 이름은 다음과 같은 패턴임: to_<데이터 확장자 이름 소문자> **(to가 아니라 read인 것 같다)**

* 예컨대 아래 3개의 메소드는 HTML, SQL, SPSS 파일 유형을 읽을 때 사용하는 메소드임

 * pandas.to_html

 * pandas.to_sql

 * pandas.to_spss

### HTML 파일 읽기

* 예시로 pandas.to_html 메소드를 사용해서 HTML 파일('*titanic.html*')을 읽어 보겠음.

* pandas.to_html은 HTML 데이터에 있는 테이블을 읽어와 DataFrame을 생성함

* 아래 그림은 볼러올 HTML 파일인 *titanic.html*을 웹브라우저로 봤을 때의 스크린샷임

![CSV Screenshot](https://drive.google.com/uc?export=view&id=1EJDBaZhrf0TXqbf-V4IB2pX0cfoEKyVV). 

* HTML 데이터에는 한개 이상의 테이블이 있을 수 있음. 따라서 이 메소드는 DataFrame을 담고 있는 list를 반환함

* 윗 그림을 보면 테이블이 한개 있음. 따라서 아래 코드의 pandas.to_html는 길이가 1개인 리스트를 반환함

In [ ]:
df = pd.read_html(path.join(root_path,'titanic.html'))
df[0].head() #html안의 모든 테이블을 가져올 수 있기 때문에 html을 읽으면 리스트로 반환함

Unnamed: 0  PassengerId  Survived  ...     Fare Cabin  Embarked
0           0            1         0  ...   7.2500  None         S
1           1            2         1  ...  71.2833   C85         C
2           2            3         1  ...   7.9250  None         S
3           3            4         1  ...  53.1000  C123         S
4           4            5         0  ...   8.0500  None         S

[5 rows x 10 columns]

## 폴더 안의 모든 데이터 파일 읽기

* 데이터 분석 작업을 하다보면 폴더에 있는 특정 유형의 모든 데이터 파일을 DataFrame으로 읽어와 리스트로 만드는 작업을 하는 경우가 많음

* 아래 코드는 '.../splits/ '폴더 안에 있는 모든 CSV 파일을 DataFrame으로 읽어와 리스트로 만드는 코드임

* [os.listdir](https://docs.python.org/3/library/os.html) 메소드는 첫주차에 "자동 파일 탐색"에서 다른 내용임. 기억이 나지 않는다면 [첫주차 colab](https://colab.research.google.com/drive/1-28UPcDO68IGnwZ0SeuG_jnvMQ--oL6E?usp=sharing) 참조

In [ ]:
import os 

directory = os.path.join(root_path,'splits')

df_list = []

for file in os.listdir(directory):
  file_path = os.path.join(directory, file)
  if os.path.isfile(file_path) and '.csv' in file: 
    df = pd.read_csv(file_path)
    df_list.append(df)

* 아래 코드는 반환된 list에서의 특정 index에 있는 DateFrame의 첫 데이터 5개을 보여주는 코드임. 

In [ ]:
import ipywidgets as widgets

print('리스트 길이 = {:d}'.format(len(df_list)))
print('보고 싶은 index 번호를 선택한 후 다음 코드를 수행하시오.')
slider = widgets.IntSlider(value=0, max=(len(df_list)-1))
display(slider)

리스트 길이 = 36
보고 싶은 index 번호를 선택한 후 다음 코드를 수행하시오.


IntSlider(value=0, max=35)

In [ ]:
df_list[slider.value].head()

PassengerId Embarked
0            3        S
1            9        S
2           11        S
3           15        S
4           19        S

# 데이터 병합

* 보통 데이터 수집 단계에 수집되는 데이터들은 다양함. 수집된 모든 데이터 파일을 병합하여 한개 데이터로 만드는 작업이 필요함

*  아래 DataFrame df1, df2을 사용해서 데이터 병합 방법에 대해 배우겠음

In [ ]:
import pandas as pd

df1 = pd.DataFrame({'A': ['df1_A0', 'df1_A1', 'df1_A2'],
                    'B': ['df1_B0', 'df1_B1', 'df1_B2'],
                    'C': ['df1_C0', 'df1_C1', 'df1_C2']},
                   index=[0, 1, 2])

df2 = pd.DataFrame({'E': ['df2_E4', 'df2_E3', 'df2_E2'],
                    'D': ['df2_D4', 'df2_D3', 'df2_D2'],
                    'C': ['df1_C4', 'df1_C3', 'df1_C2']},
                   index=[4, 3, 2])

## pandas.concat

* [Pandas.concat](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html)으로 여러 DataFrame을 한개로 병합하는 방법에 대해 배우겠음

* 데이터를 병합하는 방식은 (1) index기준 병합과 (2) column 기준 병합으로 나누어짐

* Index기준 병합은 index값 중복을 고려하지 않고 병합하는 방법으로 중복되는 column 이름이 있으면 해당되는 column의 모든 column 데이터들은 한개의 column으로 병함됨

* Column 기준 병합은 column 이름 중복을 고려하지 않고 병합하는 방법으로 중복되는 index값이 있으면 해당되는 모든 index의 row 데이터들은 한개의 row로 병함됨


### Index기준으로 데이터 병합
![CSV Screenshot](https://drive.google.com/uc?export=view&id=1ydyUjNJq7vnFBL2n_XRvC7flI-YN_-ZX) 


* Index 기준으로 여러 DataFrame을 한개로 병합하는 방법에 대해 먼저 배우겠음

* 위의 그림과 같이 Index기준 병합은 index값 중복을 고려하지 않고 병합함

* 만약 중복되는 column 이름이 있다면 해당되는 column의 모든 column 데이터들은 한개의 column으로 병함됨

* 그림에서 보면 주황색 row 두개는 병합 후에도 그래도 있지만, 빨간색 column 두개는 한 column으로 병합되었음

* 아래 코드는 Pandas.concat에 입력값으로 넣어진 DataFrame 리스트를 index기준으로 병합을 함

* 병합된 DataFrame (i.e., result)를 보면 df1과 df2의 column name = C가 한개의 column으로 병합된 것을 알 수 있음

In [ ]:
result = pd.concat([df1, df2])
result

A       B       C       E       D
0  df1_A0  df1_B0  df1_C0     NaN     NaN
1  df1_A1  df1_B1  df1_C1     NaN     NaN
2  df1_A2  df1_B2  df1_C2     NaN     NaN
4     NaN     NaN  df1_C4  df2_E4  df2_D4
3     NaN     NaN  df1_C3  df2_E3  df2_D3
2     NaN     NaN  df1_C2  df2_E2  df2_D2

* 위의 코드는 Pandas.concat의 axis 값을 '0'으로(axis=0) 했을 때와 병합 결과가 같음

* axis 값은 병합 시 사용하는 기준을 설정할 때 사용하는 입력값임. 0은 index 기존 병합을 나타내면 1은 column 기준 병합을 나타냄

* axis 값의 기본값이 0이기에 index 기존 병합을 고려할 때는 axis 값을 생략할 수 있음

In [ ]:
result = pd.concat([df1, df2], axis=0, sort=False) #axis가 0이라 index기준 병합이 됨
result

A       B       C       E       D
0  df1_A0  df1_B0  df1_C0     NaN     NaN
1  df1_A1  df1_B1  df1_C1     NaN     NaN
2  df1_A2  df1_B2  df1_C2     NaN     NaN
4     NaN     NaN  df1_C4  df2_E4  df2_D4
3     NaN     NaN  df1_C3  df2_E3  df2_D3
2     NaN     NaN  df1_C2  df2_E2  df2_D2

#### Column 정렬

* 다음 코드는 sort값으로 True로 설정하여 column을 column name에 따라 정력함
* 위의 코드 결과(column name: A, B, C, E, D)와 비교하여 column이 A, B, C, D, E 순으로 정열되어 있음을 할 수 있음

In [ ]:
result = pd.concat([df1, df2], sort=True) #column이 알파벳 순으로 정렬 sort의 기본값은 false
result

A       B       C       D       E
0  df1_A0  df1_B0  df1_C0     NaN     NaN
1  df1_A1  df1_B1  df1_C1     NaN     NaN
2  df1_A2  df1_B2  df1_C2     NaN     NaN
4     NaN     NaN  df1_C4  df2_D4  df2_E4
3     NaN     NaN  df1_C3  df2_D3  df2_E3
2     NaN     NaN  df1_C2  df2_D2  df2_E2

#### Index 재설정
* 아래 코드는 ignore_index값을 True로(ignore_index=True) 설정하여 병합 후 index 값을 [0 ~ n-1]으로 자동 재설정함

* 위의 코드 결과와 비교하여 병합된 DataFrame의 index값이 0~5로 되어 있는 것을 할 수 있음

In [ ]:
result = pd.concat([df1, df2], ignore_index=True) #index를 순서대로
result

A       B       C       E       D
0  df1_A0  df1_B0  df1_C0     NaN     NaN
1  df1_A1  df1_B1  df1_C1     NaN     NaN
2  df1_A2  df1_B2  df1_C2     NaN     NaN
3     NaN     NaN  df1_C4  df2_E4  df2_D4
4     NaN     NaN  df1_C3  df2_E3  df2_D3
5     NaN     NaN  df1_C2  df2_E2  df2_D2

#### 병합 전 DataFrame 정보 표시

* 아래 코드는 병합 시 사용된 DataFrame의 정보는 나타내는 key값을 리스트로 넣어 병합 후 index에 key값을 나타내도록 함

* 결과를 보면 key = ['df1', 'df2']에 넣어준 값들이 index에 표시된 것을 알 수 있음

In [ ]:
result = pd.concat([df1, df2], keys=['df1', 'df2'])
result

A       B       C       E       D
df1 0  df1_A0  df1_B0  df1_C0     NaN     NaN
    1  df1_A1  df1_B1  df1_C1     NaN     NaN
    2  df1_A2  df1_B2  df1_C2     NaN     NaN
df2 4     NaN     NaN  df1_C4  df2_E4  df2_D4
    3     NaN     NaN  df1_C3  df2_E3  df2_D3
    2     NaN     NaN  df1_C2  df2_E2  df2_D2

#### 중복되는 Column만 남기기
* 아래 코드는 join갑을 inner로 설절하여(join='inner') 입력된 DataFrame을 병합할 때 중복되는 column만 병합하여 반환함

* 결과를 보면 중복되는 column name = C만 반환된 것을 알 수 있음 


In [ ]:
result = pd.concat([df1, df2], join='inner') #중복되는 값만 나옴
result

C
0  df1_C0
1  df1_C1
2  df1_C2
4  df1_C4
3  df1_C3
2  df1_C2

### Column기준으로 데이터 병합
![CSV Screenshot](https://drive.google.com/uc?export=view&id=1H83ISWM1w1xVJf0SeD3FD4nkmhYTZabo) 

* 다음으로 column 기준으로 여러 DataFrame을 한개로 병합하는 방법에 대해 배우겠음

* 위의 그림과 같이 column기준 병합은 column name 중복을 고려하지 않고 병합함

* 만약 중복되는 index가 있다면 해당되는 row 데이터들은 한개의 row로 병함됨

* 그림에서 보면 빨간색 column 두개는 병합 후에도 그래도 있지만, 주황색 row 두개는 한 row로 병합되었음

* 아래 코드는 Pandas.concat에 axis값을 1로 설정하여(axis=1) 입력된 DataFrame 리스트를 column기준으로 병합을 함

* 병합된 DataFrame (i.e., result)를 보면 df1과 df2의 index = 2가 한개의 row로 병합된 것을 알 수 있음



In [ ]:
result = pd.concat([df1, df2], axis=1)
result

A       B       C       E       D       C
0  df1_A0  df1_B0  df1_C0     NaN     NaN     NaN
1  df1_A1  df1_B1  df1_C1     NaN     NaN     NaN
2  df1_A2  df1_B2  df1_C2  df2_E2  df2_D2  df1_C2
3     NaN     NaN     NaN  df2_E3  df2_D3  df1_C3
4     NaN     NaN     NaN  df2_E4  df2_D4  df1_C4

#### 병합 전 DataFrame정보 남기기

In [ ]:
result = pd.concat([df1,df2], keys=['df1', 'df2'], axis=1)
result

df1                     df2                
        A       B       C       E       D       C
0  df1_A0  df1_B0  df1_C0     NaN     NaN     NaN
1  df1_A1  df1_B1  df1_C1     NaN     NaN     NaN
2  df1_A2  df1_B2  df1_C2  df2_E2  df2_D2  df1_C2
3     NaN     NaN     NaN  df2_E3  df2_D3  df1_C3
4     NaN     NaN     NaN  df2_E4  df2_D4  df1_C4

#### 중복되는 row만 남기기

* 아래코드는 join 값을 inner로 설정하여(join='inner') 중복되는 row만 반환함

* 위의 코드 결과와 비교하면 index = 2 row만 반환된 것을 알 수 있음

In [ ]:
result = pd.concat([df1, df2], axis=1, join='inner')
result

A       B       C       E       D       C
2  df1_A2  df1_B2  df1_C2  df2_E2  df2_D2  df1_C2

#### 중복되는 row외 다른 특정 DataFrame에 속한 row도 남기기

In [ ]:
result = pd.concat([df1, df2], axis=1).reindex(df1.index)
result

A       B       C       E       D       C
0  df1_A0  df1_B0  df1_C0     NaN     NaN     NaN
1  df1_A1  df1_B1  df1_C1     NaN     NaN     NaN
2  df1_A2  df1_B2  df1_C2  df2_E2  df2_D2  df1_C2

In [ ]:
result = pd.concat([df1, df2], axis=1).reindex(df2.index)
result

A       B       C       E       D       C
4     NaN     NaN     NaN  df2_E4  df2_D4  df1_C4
3     NaN     NaN     NaN  df2_E3  df2_D3  df1_C3
2  df1_A2  df1_B2  df1_C2  df2_E2  df2_D2  df1_C2

* 윗 코드는 아래 코드와 같음

In [ ]:
result = pd.concat([df1, df2], axis=1)
result = result.reindex(df1.index)
result

A       B       C       E       D       C
0  df1_A0  df1_B0  df1_C0     NaN     NaN     NaN
1  df1_A1  df1_B1  df1_C1     NaN     NaN     NaN
2  df1_A2  df1_B2  df1_C2  df2_E2  df2_D2  df1_C2

## DataFrame.append

* [DataFrame.append](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.append.html)로 여러 DataFrame을 한개로 병합하는 방법에 대해 배우겠음

* 정확하게는 DataFrame.append는 병합이 아니고 특정 DataFrame에 다른 DataFrame 값을 **추가**하는 방식임

* 아래 DataFrame df1와 df2을 활용하여 DataFrame.append 사용법을 배우겠음


In [ ]:
import pandas as pd

df1 = pd.DataFrame({'A': ['df1_A0', 'df1_A1', 'df1_A2'],
                    'B': ['df1_B0', 'df1_B1', 'df1_B2'],
                    'C': ['df1_C0', 'df1_C1', 'df1_C2']},
                   index=[0, 1, 2])

df2 = pd.DataFrame({'E': ['df2_E4', 'df2_E3', 'df2_E2'],
                    'D': ['df2_D4', 'df2_D3', 'df2_D2'],
                    'C': ['df1_C4', 'df1_C3', 'df1_C2']},
                   index=[4, 3, 2])

* 아래 코드는 df1에 df2의 데이터를 추가한 후 반환함

In [ ]:
result = df1.append(df2) #index기준 병합과 같아지네
result

A       B       C       E       D
0  df1_A0  df1_B0  df1_C0     NaN     NaN
1  df1_A1  df1_B1  df1_C1     NaN     NaN
2  df1_A2  df1_B2  df1_C2     NaN     NaN
4     NaN     NaN  df1_C4  df2_E4  df2_D4
3     NaN     NaN  df1_C3  df2_E3  df2_D3
2     NaN     NaN  df1_C2  df2_E2  df2_D2

* DataFrame.concate를 사용하여 index기준 병합을 한 것과 차이가 없음

In [ ]:
result = pd.concat([df1, df2])
result

A       B       C       E       D
0  df1_A0  df1_B0  df1_C0     NaN     NaN
1  df1_A1  df1_B1  df1_C1     NaN     NaN
2  df1_A2  df1_B2  df1_C2     NaN     NaN
4     NaN     NaN  df1_C4  df2_E4  df2_D4
3     NaN     NaN  df1_C3  df2_E3  df2_D3
2     NaN     NaN  df1_C2  df2_E2  df2_D2

#### 여러 DataFrame 첨부하기

* 특정 DataFrame에 여러 DataFrame 데이터를 동시에 추가할 수 도 있음

* DataFrame df3을 추가하여 df2, df3 데이터를 df1에 추가하여 보겠음

In [ ]:
df3 = pd.DataFrame({'F': ['df2_F5', 'df2_F6', 'df2_F7'],
                    'G': ['df2_G5', 'df2_G6', 'df2_G7'],
                    'H': ['df1_H5', 'df1_H6', 'df1_H7']},
                   index=[5, 6, 7])

In [ ]:
result = df1.append([df2, df3])
result

A       B       C       E       D       F       G       H
0  df1_A0  df1_B0  df1_C0     NaN     NaN     NaN     NaN     NaN
1  df1_A1  df1_B1  df1_C1     NaN     NaN     NaN     NaN     NaN
2  df1_A2  df1_B2  df1_C2     NaN     NaN     NaN     NaN     NaN
4     NaN     NaN  df1_C4  df2_E4  df2_D4     NaN     NaN     NaN
3     NaN     NaN  df1_C3  df2_E3  df2_D3     NaN     NaN     NaN
2     NaN     NaN  df1_C2  df2_E2  df2_D2     NaN     NaN     NaN
5     NaN     NaN     NaN     NaN     NaN  df2_F5  df2_G5  df1_H5
6     NaN     NaN     NaN     NaN     NaN  df2_F6  df2_G6  df1_H6
7     NaN     NaN     NaN     NaN     NaN  df2_F7  df2_G7  df1_H7

#### Column 정렬

* 다음 코드는 sort값으로 True로 설정하여 column을 column name에 따라 정력함
* 위의 코드 결과(column name: A, B, C, E, D)와 비교하여 column이 A, B, C, D, E 순으로 정열되어 있음을 할 수 있음

In [ ]:
result = df1.append(df2, sort=True)
result

A       B       C       D       E
0  df1_A0  df1_B0  df1_C0     NaN     NaN
1  df1_A1  df1_B1  df1_C1     NaN     NaN
2  df1_A2  df1_B2  df1_C2     NaN     NaN
4     NaN     NaN  df1_C4  df2_D4  df2_E4
3     NaN     NaN  df1_C3  df2_D3  df2_E3
2     NaN     NaN  df1_C2  df2_D2  df2_E2

#### Index 재설정

* 아래 코드는 ignore_index값을 True로(ignore_index=True) 설정하여 병합 후 index 값을 [0 ~ n-1]으로 자동 재설정함

* 위의 코드 결과와 비교하여 병합된 DataFrame의 index값이 0~5로 되어 있는 것을 할 수 있음

In [ ]:
 result = df1.append(df2, ignore_index=True)
 result

A       B       C       E       D
0  df1_A0  df1_B0  df1_C0     NaN     NaN
1  df1_A1  df1_B1  df1_C1     NaN     NaN
2  df1_A2  df1_B2  df1_C2     NaN     NaN
3     NaN     NaN  df1_C4  df2_E4  df2_D4
4     NaN     NaN  df1_C3  df2_E3  df2_D3
5     NaN     NaN  df1_C2  df2_E2  df2_D2

## 폴더 안의 모든 데이터 파일 병합하기

* 폴더 안의 모든 데이터 파일 읽기 section에서 splits 폴더(i.e., '*.../splites/*') 안에 있는 모든 CSV 파일을 DataFrame으로 읽어와 리스트(list)로 만드는 것을 배웠음

* 이번에는 CSV파일을 읽어온 후 병합하여 아래 테이블 같은 형태의 DataFrame을 만들어 보겠음

index| PassengerId| Survived| Pclass| Sex| Age| Ticket| Fare| Cabin| Embarked
:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:| 
0|1|0|3|male|22.0|A/5 21171|7.2500|NaN|S
1|2|1|1|female|38.0|PC 17599|71.2833|C85|C
2|3|1|3|female|26.0|STON/O2. 3101282|7.9250|NaN|S
3|4|1|1|female|35.0|113803|53.1000|C123|S
4|5|0|3|male|35.0|373450|8.0500|NaN|S

* 윗 테이블의 'titanic.csv'로 생성한 DataFrame 형태와 같음

In [ ]:
import pandas as pd
from os import path

df = pd.read_csv(path.join(root_path,'titanic.csv'))
df.head()

PassengerId  Survived  Pclass  ...     Fare  Cabin Embarked
0            1         0       3  ...   7.2500    NaN        S
1            2         1       1  ...  71.2833    C85        C
2            3         1       3  ...   7.9250    NaN        S
3            4         1       1  ...  53.1000   C123        S
4            5         0       3  ...   8.0500    NaN        S

[5 rows x 9 columns]

* splits 폴더에 있는 파일은 아래와 그림과 같음

![CSV Screenshot](https://drive.google.com/uc?export=view&id=1qb1Nh28ywoTL1mGgSDyXDcOs8pgMAyVu) 

*  개별 파일은 특정 Pclass의 특정 성별을 가진 탑승객의 {Survived, Age, Ticket, Fare, Cabin, 	Embarked}의 정보 중 한개를 담고 있음

* 예컨대 'titanic_Age_class1_female.csv' 파일을 보면 Class1에 탑승한 여성 승객의 PassengerId와 Age정보가 나와 있음

In [ ]:
import pandas as pd
from os import path

directory = os.path.join(root_path,'splits')

df = pd.read_csv(path.join(directory,'titanic_Age_class1_female.csv'))
df.head()

PassengerId   Age
0            2  38.0
1            4  35.0
2           12  58.0
3           32   NaN
4           53  49.0

* 먼저 동일 종류(e.g., Age)의 정보를 담고 있는 파일들을 index 기준 병합 한 후, 다음으로 병합된 DataFrame을 column 기준 병합 하겠음 

In [ ]:
import os 

directory = os.path.join(root_path,'splits')

info_types = ['Survived', 'Age', 'Ticket', 'Fare', 'Cabin', 'Embarked']
info_type_dfs = []

for info_type in info_types:
  df_list = []
  for file in os.listdir(directory):
    file_path = os.path.join(directory, file)
    if os.path.isfile(file_path) and '.csv' in file and info_type in file: 
      df = pd.read_csv(file_path)
      if 'Survived' in file:
        df['sex'] = 'female' if 'female' in file else 'male'
        df['Pclass'] = 1 if 'class1' in file else 2 if 'class2' in file else 3
     
      df_list.append(df)

  df = pd.concat(df_list, ignore_index=True)
  

  df = df.set_index('PassengerId')
  info_type_dfs.append(df)

In [ ]:
info_type_dfs[0].head()

Survived   sex  Pclass
PassengerId                        
1                   0  male       3
5                   0  male       3
6                   0  male       3
8                   0  male       3
13                  0  male       3

In [ ]:
final_df = pd.concat(info_type_dfs, axis=1)
final_df = final_df.reset_index()
final_df

PassengerId  Survived     sex  ...     Fare  Cabin Embarked
0              1         0    male  ...   7.2500    NaN        S
1              2         1  female  ...  71.2833    C85        C
2              3         1  female  ...   7.9250    NaN        S
3              4         1  female  ...  53.1000   C123        S
4              5         0    male  ...   8.0500    NaN        S
..           ...       ...     ...  ...      ...    ...      ...
886          887         0    male  ...  13.0000    NaN        S
887          888         1  female  ...  30.0000    B42        S
888          889         0  female  ...  23.4500    NaN        S
889          890         1    male  ...  30.0000   C148        C
890          891         0    male  ...   7.7500    NaN        Q

[891 rows x 9 columns]

* 아래 코드는 'titanic.csv'로 생성한 DataFrame를 보여주고 있음. Column name 순서는 다르지만 윗 final_df의 정보와 같음을 알 수 있음

In [ ]:
import pandas as pd
from os import path

df = pd.read_csv(path.join(root_path,'titanic.csv'))
df

PassengerId  Survived  Pclass  ...     Fare  Cabin Embarked
0              1         0       3  ...   7.2500    NaN        S
1              2         1       1  ...  71.2833    C85        C
2              3         1       3  ...   7.9250    NaN        S
3              4         1       1  ...  53.1000   C123        S
4              5         0       3  ...   8.0500    NaN        S
..           ...       ...     ...  ...      ...    ...      ...
886          887         0       2  ...  13.0000    NaN        S
887          888         1       1  ...  30.0000    B42        S
888          889         0       3  ...  23.4500    NaN        S
889          890         1       1  ...  30.0000   C148        C
890          891         0       3  ...   7.7500    NaN        Q

[891 rows x 9 columns]

# 데이터 삭제

* DataFrame의 특정 Row나 Column을 삭제할 때는 [DataFrame.drop](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop.html) 메소드를 사용하면 됨

* 삭제하고자 하는 데이터의 타입(Row 또는 Column)에 따라 index, columns 값을 넣어주면 됨

* DataFrame df1, df2, df3를 병합한 df를 활용하여 데이터 삭제하는 것에 대해 배우겠음 


In [ ]:
import pandas as pd
import numpy as np

df1 = pd.DataFrame({'A': ['df1_A0', 'df1_A1', 'df1_A2'],
                    'B': ['df1_B0', 'df1_B1', 'df1_B2'],
                    'C': ['df1_C0', 'df1_C1', 'df1_C2']},
                   index=[0, 1, 2])
df2 = pd.DataFrame({'E': ['df2_E4', 'df2_E3', 'df2_E2'],
                    'D': ['df2_D4', 'df2_D3', 'df2_D2'],
                    'C': ['df1_C4', 'df1_C3', 'df1_C2']},
                   index=[4, 3, 2])
df3 = pd.DataFrame({'F': [np.nan, np.nan, np.nan],
                    'G': [np.nan, 'df3_G6', 'df3_G7'],
                    'C': ['df3_C6', 'df3_C6', 'df3_C7']},
                   index=[5, 6, 7])

def get_df():
  return pd.concat([df1, df2, df3], sort=True, ignore_index=True)

df = get_df()
df

A       B       C       D       E   F       G
0  df1_A0  df1_B0  df1_C0     NaN     NaN NaN     NaN
1  df1_A1  df1_B1  df1_C1     NaN     NaN NaN     NaN
2  df1_A2  df1_B2  df1_C2     NaN     NaN NaN     NaN
3     NaN     NaN  df1_C4  df2_D4  df2_E4 NaN     NaN
4     NaN     NaN  df1_C3  df2_D3  df2_E3 NaN     NaN
5     NaN     NaN  df1_C2  df2_D2  df2_E2 NaN     NaN
6     NaN     NaN  df3_C6     NaN     NaN NaN     NaN
7     NaN     NaN  df3_C6     NaN     NaN NaN  df3_G6
8     NaN     NaN  df3_C7     NaN     NaN NaN  df3_G7

## Column 데이터 삭제
* 삭제하고자 하는 column name을 리스트(e.g., ['C', 'D', 'E', 'G'])로 만들어 DataFrame.drop 메소드의 columns 값으로 설정하면 됨

* drop 메소드는 columns 값에 따라 column이 삭제된 DataFrame이 반환함

* 아래 코드는 drop 메소드의 columns 값을 ['C', 'D', 'E', 'G']로 설정하여 해당 column이 삭제된 DataFrame을 반환하는 코드임

* 결과를 보면 ['C', 'D', 'E', 'G'] column이 없는 것을 알 수 있음

In [ ]:
df = get_df()
df = df.drop(columns=['C', 'D', 'E', 'G'])
df

A       B   F
0  df1_A0  df1_B0 NaN
1  df1_A1  df1_B1 NaN
2  df1_A2  df1_B2 NaN
3     NaN     NaN NaN
4     NaN     NaN NaN
5     NaN     NaN NaN
6     NaN     NaN NaN
7     NaN     NaN NaN
8     NaN     NaN NaN

## Row 데이터 삭제

* 삭제하고자 하는 row의 index을 리스트(e.g., [1, 3, 4, 8])로 만들어 DataFrame.drop 메소드의 index 값으로 설정하면 해당 row가 삭제된 DataFrame이 반환함

* 아래 코드는 drop 메소드의 index 값을 [1, 3, 4, 8]로 설정하여 해당 row가 삭제된 DataFrame을 반환하는 코드임

* 결과를 보면 [1, 3, 4, 8]를 index로 가지는 row가 없는 것을 알 수 있음

In [ ]:
df = get_df()
df = df.drop(index=[1, 3, 4, 8])
df

A       B       C       D       E   F       G
0  df1_A0  df1_B0  df1_C0     NaN     NaN NaN     NaN
2  df1_A2  df1_B2  df1_C2     NaN     NaN NaN     NaN
5     NaN     NaN  df1_C2  df2_D2  df2_E2 NaN     NaN
6     NaN     NaN  df3_C6     NaN     NaN NaN     NaN
7     NaN     NaN  df3_C6     NaN     NaN NaN  df3_G6

## 빈칸(NaN) 데이터 삭제
* [DataFrame.dropna](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dropna.html) 메소드는 axis 값(i.e., 0 또는 1)과 how 값(i.e., any 또는 all)에 따라 빈칸(NaN)을 가진 axis를 삭제함

* DataFrame.dropna의 입력값 중 axis 값은 row(axis=0) 또는 column(axis=1)을 중 어떤 axis를 삭제할 것인지 설정하는 것임.

* 입력값 중 how 값은 빈칸(NaN)이 어떤 형태로 있을 때 삭제할 것인지 설정하는 것임

 * how 값을 any로 설정하면(how='any') 빈칸(NaN)이 한개하도 있으면 해당 axis를 삭제함

 * how 값을 all로 설정하면(how='all') 해당 axis의 모든 데이터가 빈칸(NaN)이면 해당 axis를 삭제함

* 아래 코드는 column의 모든 데이터가 빈칸일때 해당 column를 삭제함. 

* 결과를 보면 column name = F가 없을 것을 알 수 있음

In [ ]:
df = get_df()
df = df.dropna(axis=1, how='all')
df

A       B       C       D       E       G
0  df1_A0  df1_B0  df1_C0     NaN     NaN     NaN
1  df1_A1  df1_B1  df1_C1     NaN     NaN     NaN
2  df1_A2  df1_B2  df1_C2     NaN     NaN     NaN
3     NaN     NaN  df1_C4  df2_D4  df2_E4     NaN
4     NaN     NaN  df1_C3  df2_D3  df2_E3     NaN
5     NaN     NaN  df1_C2  df2_D2  df2_E2     NaN
6     NaN     NaN  df3_C6     NaN     NaN     NaN
7     NaN     NaN  df3_C6     NaN     NaN  df3_G6
8     NaN     NaN  df3_C7     NaN     NaN  df3_G7

* 아래 코드는 row의 데이터에 빈칸이 한개라도 있으면 삭제함

* 결과를 보면 모든 row가 삭제됨


In [ ]:
df = get_df()
df = df.dropna(axis=0, how='any')
df

Empty DataFrame
Columns: [A, B, C, D, E, F, G]
Index: []